File: src/backend/finance-service/Dtos/Responses/BudgetResponse.cs
namespace FinanceService.Dtos.Responses;

public class BudgetResponse
{
    public long Id { get; set; }
    public int FiscalYear { get; set; }
    public decimal? AnnualTarget { get; set; }
    public decimal? ApprovedAmount { get; set; }
    public decimal? YtdActuals { get; set; }
    public decimal? LatestForecast { get; set; }
    public bool IsLocked { get; set; }
    public DateTime? LockedAt { get; set; }
    public string? Owner { get; set; }
    public string? Notes { get; set; }
    public DateTime CreatedAt { get; set; }
    public DateTime UpdatedAt { get; set; }
    public IReadOnlyList<BudgetStageResponse> Stages { get; set; } = Array.Empty<BudgetStageResponse>();
    public IReadOnlyList<BudgetSnapshotResponse> Snapshots { get; set; } = Array.Empty<BudgetSnapshotResponse>();
}

public class BudgetStageResponse
{
    public long Id { get; set; }
    public string Stage { get; set; } = string.Empty;
    public string Label { get; set; } = string.Empty;
    public string State { get; set; } = string.Empty;
    public string? Owner { get; set; }
    public string? Notes { get; set; }
    public DateTime? StartedAt { get; set; }
    public DateTime? CompletedAt { get; set; }
}

public class BudgetSnapshotResponse
{
    public long Id { get; set; }
    public string Month { get; set; } = string.Empty;
    public decimal? BudgetAmount { get; set; }
    public decimal? ActualAmount { get; set; }
    public decimal? ForecastAmount { get; set; }
    public decimal? Variance { get; set; }
    public string? Notes { get; set; }
}

File: src/backend/report-service/src/main/java/com/example/report_service/clients/finance/dto/BudgetResponse.java
package com.example.report_service.clients.finance.dto;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonProperty;
import java.math.BigDecimal;
import java.time.Instant;
import java.util.List;

@JsonIgnoreProperties(ignoreUnknown = true)
public record BudgetResponse(
        Long id,
        Integer fiscalYear,
        BigDecimal annualTarget,
        BigDecimal approvedAmount,
        BigDecimal ytdActuals,
        BigDecimal latestForecast,
        Boolean isLocked,
        Instant lockedAt,
        String owner,
        String notes,
        Instant createdAt,
        Instant updatedAt,
        List<BudgetStageResponse> stages,
        List<BudgetSnapshotResponse> snapshots) {

    @JsonCreator
    public BudgetResponse(
            @JsonProperty("id") Long id,
            @JsonProperty("fiscalYear") Integer fiscalYear,
            @JsonProperty("annualTarget") BigDecimal annualTarget,
            @JsonProperty("approvedAmount") BigDecimal approvedAmount,
            @JsonProperty("ytdActuals") BigDecimal ytdActuals,
            @JsonProperty("latestForecast") BigDecimal latestForecast,
            @JsonProperty("isLocked") Boolean isLocked,
            @JsonProperty("lockedAt") Instant lockedAt,
            @JsonProperty("owner") String owner,
            @JsonProperty("notes") String notes,
            @JsonProperty("createdAt") Instant createdAt,
            @JsonProperty("updatedAt") Instant updatedAt,
            @JsonProperty("stages") List<BudgetStageResponse> stages,
            @JsonProperty("snapshots") List<BudgetSnapshotResponse> snapshots) {
        this.id = id;
        this.fiscalYear = fiscalYear;
        this.annualTarget = annualTarget;
        this.approvedAmount = approvedAmount;
        this.ytdActuals = ytdActuals;
        this.latestForecast = latestForecast;
        this.isLocked = isLocked;
        this.lockedAt = lockedAt;
        this.owner = owner;
        this.notes = notes;
        this.createdAt = createdAt;
        this.updatedAt = updatedAt;
        this.stages = stages == null ? List.of() : List.copyOf(stages);
        this.snapshots = snapshots == null ? List.of() : List.copyOf(snapshots);
    }
}
